### PR TITLE
commit: do not change container labels on commit

### LINKED ISSUE #29547: When commiting an image with changes the running container also changes
**Description**

When commiting an image with changes (ie labels) the running container's labels are also changed.

**Steps to reproduce the issue:**
1. ```docker run -d --label some=label centos:7```
2. ```docker commit --change "LABEL some=label2" ${c_id} my/image```
3. ```docker inspect ${c_id}```

**Describe the results you received:**

The running container's labels are also changed.

**Describe the results you expected:**

The running container does not change.

**Output of `docker version`:**

```
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 23:26:11 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.12.4
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   1564f02
 Built:        Tue Dec 13 02:47:26 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 28
 Running: 26
 Paused: 0
 Stopped: 2
Images: 30
Server Version: 1.12.4
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 404
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.38-boot2docker
Operating System: Boot2Docker 1.12.4 (TCL 7.2); HEAD : d0b8fd8 - Tue Dec 13 18:21:26 UTC 2016
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 3.858 GiB
Name: bats
ID: 2FFF:SLRA:BJOP:L6Z2:2UQU:VR3J:DGJ2:UOFR:SH27:UCW4:I26R:PKFJ
Docker Root Dir: /mnt/sda1/var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 215
 Goroutines: 177
 System Time: 2016-12-19T16:31:03.949019346Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
Insecure Registries:
 127.0.0.0/8
```

### PR DESCRIPTION
**Pull request by LK4D4**
Fix #29547
ping @cpuguy83

### CODE REVIEW

File: daemon/commit.go [hunk 1]
```diff
@@ -65,11 +65,10 @@ func merge(userConf, imageConf *containertypes.Config) error {
 	if userConf.Labels == nil {
 		userConf.Labels = map[string]string{}
 	}
-	if imageConf.Labels != nil {
-		for l := range userConf.Labels {
-			imageConf.Labels[l] = userConf.Labels[l]
+	for l, v := range imageConf.Labels {
+		if _, ok := userConf.Labels[l]; !ok {
```
Comment (cpuguy83) [inline daemon/commit.go]: What about when the key does exist? Shouldn't we still set it?

**Comment by tonistiigi** [comment]
Why pass the `imageConf` as a pointer at all? As a caller I wouldn't know if I need to make copy of it first.

**Comment by cpuguy83** [comment]
Even if it wasn't a pointer, we'd still be accessing the container's actual labels in the `merge` function.

File: daemon/commit.go (hunk 1 above)
Comment (LK4D4) [inline daemon/commit.go]: Have no idea, I tried to mimic old behavior :/

**Comment by LK4D4** [comment]
@tonistiigi Why not? it's big structure and `merge` is just helper. Also, yeah, it would require deep copy :/

**Comment by tonistiigi** [comment]
@LK4D4 Not that important. Just if I see a pointer passed I must assume that a function is doing that because it wants to modify the inputs. This is not a performance-critical code path where a copy would have any effect. But yes, it is a private method and doesn't even fix this specific issue with labels.

**Review by dnephin — APPROVED** [review]
LGTM

I think this makes sense. The userConf labels take precedence over the imageConf labels, as it was before.

**Review by tonistiigi — APPROVED** [review]

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
ping @vieux This probably needs to go in 1.13

**Pull request merged**